lib/auth.js

```diff
@@ -6,10 +6,59 @@ const _ = require('lodash'),
   references = require('./services/references'),
   session = require('express-session'),
   flash = require('express-flash'),
+  responses = require('./responses'),
   handlebars = require('handlebars'),
   basicAuth = require('basic-auth'),
   fs = require('fs'),
-  path = require('path');
+  path = require('path'),
+  AUTH_LEVELS_MAP = {
+    ADMIN: 'admin',
+    WRITE: 'write',
+    READ: 'read'
+  },
+  AUTH_LEVEL_VALUES = {
+    admin: 0,
+    write: 1,
+    read: 2
+  };
+
+/**
+ * Check the auth level to see if a user
+ * has sufficient permissions
+ *
+ * @param  {String} userLevel
+ * @param  {String} requiredLevel
+ * @return {Boolean}
+ */
+function checkAuthLevel(userLevel, requiredLevel) {
+  // User has to have an auth level set
+  if (!userLevel) {
+    throw new Error('User does not have an authentication level set');
+  }
+
+  // Compare levels, the lover you go the more permissions you have
+  return AUTH_LEVEL_VALUES[userLevel] <= AUTH_LEVEL_VALUES[requiredLevel];
+}
+
+/**
+ * Get the user auth level and check it against the
+ * required auth level for a route. Send an error
+ * if the user doesn't have permissions
+ *
+ * @param  {String} requiredLevel
+ * @return {Function}
+ */
+function withAuthLevel(requiredLevel) {
+  return function (req, res, next) {
+    if (checkAuthLevel(_.get(req, 'user.auth', ''), requiredLevel)) {
```

Comment:
If `user` is undefined then an error will be thrown about not being able to read `auth` of undefined. Lodash's get method lets us choose a value to pass in in case the desired property doesn't exist, so this is just insurance that we get to the `checkAuthLevel` function.